### ISSUE TITLE
Why does `pkgs.lua` resolve to lua 5.2.4 instead of a more modern version

### ISSUE THREAD

**Issue by Frontear** [NixOS/nixpkgs#323083]
This issue is inspired by https://github.com/NixOS/nixpkgs/issues/323016.

I found it quite interesting that `pkgs.lua` resolves to lua 5.2.4, and especially given that this released way back in 2011, is probably due an update. Lua 5.4.x, by contrast, has been released since 2020, with the most recent version being 5.4.7.

I know normally we hold off pushing a major bump (see python3 being on 3.11 when 3.13 is out), but Lua has been out for many years. Seems like it's about time for a rev bump.

If it's legacy compat we are concerned with, then have them changed to `pkgs.lua_5_2` (i wouldn't even mind making this pr). I think bumping the generic alias to a higher default is better for most people, new Nix/OS users, and even older users looking to adopt newer Lua packages.

Honestly I have a similar (albiet I guess less important) issue with `pkgs.python` resolving to python2 and `pkgs.python3` being the necessary output instead of the other way around `pkgs.python ==> python3` and `pkgs.python2 ==> python2`.

**Comment by eclairevoyant** [comment]
lua is not like other languages, they make breaking changes even on "minor" versions. 5.2 and 5.3 are practically different languages

**Comment by flokli** [comment]
Policy is usually that the alias points to the latest version.

> lua is not like other languages, they make breaking changes even on "minor" versions. 5.2 and 5.3 are practically different languages

Hmm, that would be an argument against having a `lua` alias pointing to the latest at all.

I agree `lua` pointing to an old version is confusing.

**Comment by Frontear** [comment]
Bumping this, I think more consensus would be really cool here. Almost every other aliased language-specific package points to the latest version, and for a repository like nixpkgs where everything is ideally kept bleeding-edge, it makes sense.

Lua is the one of (if not the only) language that isn't following this logic, and I believe its confusing for both developers who expect modern lua as they work, and packagers who bring it modern lua packages. Realistically if a package requires an older lua, it should be provided into their attributes either as `lua52` or `lua = lua52`.

**Comment by flokli** [comment]
I agree, and if specific applications require specific minor version, I'd rather see it explicitly written down. Wanna send a draft PR so we can see the impact?

**Comment by galenhuntington** [comment]
Another consideration is that some rocks in nixpkgs are so out of date that they won't work with recent Lua.  I recently tried to install `luaposix` for Lua 5.4 only to find that it's marked as broken.  The reason is that in nixpkgs, `luaposix` is still on version 34.1.1, released 2019-07-27; Lua 5.4 support was added in version 25.0, which was released 2020-07-06.  In other words, the version here is more than 5 years out of date and stuck at a version incompatible with Lua 5.4.

There are several broken packages: https://github.com/NixOS/nixpkgs/pull/399219#issuecomment-2809953708
Some may also be due to outdatedness.  It seems there is not a sufficient process for updating.

**Comment by teto** [comment]
> lua is not like other languages, they make breaking changes even on "minor" versions. 5.2 and 5.3 are practically different languages

That's why. lua is popular in embedded where once you set on a version, you dont upgrade. This can be seen in the amybe more brittle package manager than in other languages, people would just bundle the dependencies.
For instance neovim uses 5.1 with no plan to upgrade, also because luajit only supports 5.1 and is a super fast interpreter.

> Hmm, that would be an argument against having a lua alias pointing to the latest at all.

Makes sense. Not strong enough to remove the current alias IMO as it is practical as well if you just dabble with it. 

> Another consideration is that some rocks in nixpkgs are so out of date that they won't work with recent Lua. 
> ...
> There are several broken packages: https://github.com/NixOS/nixpkgs/pull/399219#issuecomment-2809953708

This needs to be nuanced with the previous points. Right now we mark as broken packages for interpreters that are not supported by the author but it's not actually "broken". Now that we have meta.problems, we could probably show a better explanation message.

>  It seems there is not a sufficient process for updating.

The updater is more robust than ever and there are regular updates these days, mostly by @khaneliman , mostly driven by neovim dependencies that target 5.1 for instance.

While I would have preferred the `lua` alias to target lua5.1 as it the most popular version of lua in nixpkgs I beleive, now that it has been updated to 5.2, we might as well target 5.4 (5.5 is too recent, too few packages support it).

I have no motivation/time to do it but can review/merge the change. Process would be:
- update alias
- run a nixpkgs-review
- compare "broken" packages against master
- fix what is relevant
- profit

**Comment by Frontear** [comment]
> Makes sense. Not strong enough to remove the current alias IMO as it is practical as well if you just dabble with it.

Could you elaborate on some of these practical benefits? I feel like it doesn't make sense to have an un-versioned alias if it doesn't conform to the Nixpkgs tradition of being the latest version. It feels counterintuitive.

**Comment by teto** [comment]
`lua` is` more discoverable than `lua5_2` and if you just want a REPL, the version wont matter much. 
Breaking backwards compatibility by removing the alias requires a clear win and here it's not the case. The original question was why `lua` doesn't point to latest version, point to 5.5 would result in lots of "broken" lua packages so not a smart choice either. 
One has to accept that ecosystems differ, haskell follows stackage which isn't  latest version either.
And as stated in my previous post, I welcome a bump of lua to lua5.4 if someone steps up and do the work.

**Comment by flokli** [comment]
We can keep a `lua` alias in `aliases.nix` (so a casual/external user will get some lua), but make other references explicit.

Then the nixpkgs tooling will at least yell if `lua` is used without being explicit about the version from inside nixpkgs.
